File: part1/fortran/mat_mul_par.f90
!*********************************************************************
! Filename:      mat_mul_par.f90
! Author:        Prasad Maddumage <mhemantha@fsu.edu>
! Created at:    Mon Jul 14 08:49:19 2014
! Modified at:   Wed Jul 16 15:41:30 2014
! Modified by:   Prasad Maddumage <mhemantha@fsu.edu>
! Version:       $Revision: 1.3 $
! Description:   Parallel matrix multiplication
!*********************************************************************
program mat_mul_par

  implicit none

!Include the MPI header file
  include 'mpif.h'

!number of rows in A, B = nr_a, nr_b
!number of columns in A, B = nc_a, nc_b
  integer, parameter :: MASTER = 0, nc_a = 4, nr_a = 4, nr_b = 4, nc_b = 4
  integer            :: n_procs, my_rank, len, ierr
  integer            :: nc_local, n2, n_el, i, j, k
  real(kind = 8)     :: t_start, t_finish
  real               :: t1, t2, a(nr_a,nc_a), b(nr_b,nc_b), c(nr_a,nc_b)
  real, allocatable  :: b_local(:,:), c_local(:,:)

!Initializing MPI
  call MPI_INIT(ierr)
  call MPI_COMM_SIZE(MPI_COMM_WORLD, n_procs, ierr)
  call MPI_COMM_RANK(MPI_COMM_WORLD, my_rank, ierr)

!Initialize matrices A and B with some data (on MASTER node)
  if (my_rank == MASTER) then
       do i = 1, nr_a
           do j = 1, nc_a
               a(i,j) = real(i + j)
            end do
       end do
       
       do i = 1, nr_b
           do j = 1, nc_b
               b(i,j) = real(i * j)
           end do
       end do
!Start counting time
       t_start = MPI_Wtime()
  end if
  
!Load balancing: each process calculate same number of elements in matrix C
!Let each process only calculate few columns of C
!Matrix multiplication: C(i,j) = {sum over all k} A(i,k) * B(k,j)
!Therefore, each process needs only few columns of matrix B
!nc_local = number of columns calculated by each process
  nc_local = ceiling(real(nc_b) / n_procs)
!nc_b >= n_procs must be true

!B_local and C_local will hold the columns needed/calculated
!by a SINGLE process
  allocate(b_local(nr_b,nc_local), c_local(nr_a,nc_local))
  b_local = 0.0; c_local = 0.0

!All processes need matrix A. Therefore, BCAST it
  n_el = nr_a * nc_a !Total number of elements in matrix A
  call MPI_BCAST(a, n_el, MPI_REAL, MASTER, MPI_COMM_WORLD, ierr)
!Send n columns each from matrix B to ALL processes (including MASTER)
  n2 = nc_local * nr_b !# of elements in n number of columns
  call MPI_SCATTER(b, n2, MPI_REAL, b_local, n2, MPI_REAL, MASTER, &
         MPI_COMM_WORLD, ierr)

!Do the matrix multiplication to calculate n columns of C
!Notice all processes do the work including MASTER
  do j = 1, nc_local
      do i = 1, nr_a
          do k = 1, nc_a
              c_local(i,j) = c_local(i,j) + a(i,k) * b_local(k,j)
          end do
      end do
  end do

!Collect the partial results to form the final matrix C on MASTER
  call MPI_GATHER(c_local, n2, MPI_REAL, c, n2, MPI_REAL, MASTER, &
         MPI_COMM_WORLD, ierr)

!Print the result as a matrix
  if (my_rank == MASTER) then
       t_finish = MPI_Wtime()
       print *, "Time elapsed for parallel computation (s) ", t_finish - t_start
      
       do i = 1, nr_a
           print *, c(i,:)
       end do
  end if

  deallocate(b_local, c_local)
  call MPI_FINALIZE(ierr)

end program mat_mul_par
  

File: part1/c/mat_mul_par.c
/*********************************************************************
 * Filename:      mat_mul_par.c
 * Author:        Prasad Maddumage <mhemantha@fsu.edu>
 * Created at:    Tue Jul 15 13:06:17 2014
 * Modified at:   Tue Oct  3 00:25:30 2017
 * Modified by:   Prasad Maddumage <mhemantha@fsu.edu>
 * Version:       $Revision: 1.2 $
 * Description:   Parallel matrix multiplication
 ********************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "mpi.h"

//number of rows in A, B = nr_a, nr_b
//number of columns in A, B = nc_a, nc_b
const int master = 0, nc_a = 4, nr_a = 4, nr_b = 4, nc_b = 4;

int main (int argc, char* argv[])
{

    int n_procs, my_rank, source, tag = 0, ierr;
    int nr_local, n2, n_el, i, j, k;
    double t_start, t_finish;
    double a[nr_a][nc_a], b[nr_b][nc_b], c[nr_a][nc_b];
    double *a_local, *c_local;

    //Initializing MPI 
    ierr = MPI_Init(&argc, &argv);

    ierr = MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
    ierr = MPI_Comm_size(MPI_COMM_WORLD, &n_procs);

    //Initialize matrices A and B with some data (on MASTER node)
    if (my_rank == master) {
      for (i = 0; i < nr_a; i++) {
	for (j = 0; j < nc_a; j++) {
	  a[i][j] = (double)((1 + i) + (1 + j));
	}
      }
      for (i = 0; i < nr_b; i++) {
	for (j = 0; j < nc_b; j++) {
	  b[i][j] = (double)((1 + i) * (1 + j));
	}
      }
      //Start counting time
      t_start = MPI_Wtime();
    }

    //Load balancing: each process calculate same number of elements in matrix C
    //C is row major. So, let each process only calculate few rows of matrix C
    //Matrix multiplication: C(i,j) = {sum over all k} A(i,k) * B(k,j)
    //Each process needs only few rows of matrix A
    //nr_local = number of rows calculated by each process
    nr_local = ceil((double)nr_a / (double)n_procs);
    //nr_a >= n_procs must be true
    
    //A_local and C_local will hold the rows needed/calculated
    //by a SINGLE process
    a_local = (double *) calloc(nr_local * nc_a, sizeof(double));
    c_local = (double *) calloc(nr_local * nc_b, sizeof(double));
    for(i = 0; i < nr_local*nc_a; i++) {
      a_local[i] = 0.0;
    }
    for(i = 0; i < nr_local*nc_b; i++) {
      c_local[i] = 0.0;
    }

    //All processes need matrix B. Therefore, BCAST it
    n_el = nr_b * nc_b; //Total number of elements in matrix A
    ierr = MPI_Bcast(b, n_el, MPI_DOUBLE, master, MPI_COMM_WORLD);
    //Send n columns each from matrix B to ALL processes (including master)
    n2 = nr_local * nc_a; //# of elements in n number of columns
    ierr = MPI_Scatter(a, n2, MPI_DOUBLE, a_local, n2, MPI_DOUBLE, master, \
		       MPI_COMM_WORLD);
    
    for (j = 0; j < nc_b; j++) {
      for (i = 0; i < nr_local; i++) {
	for (k = 0; k < nc_a; k++) {
	  c_local[i*nr_local+j] = c_local[i*nr_local+j] + \
	    a_local[i*nr_local+k] * b[k][j];
	}
      }
    }

    //Collect the partial results to form the final matrix C on master
    ierr = MPI_Gather(c_local, n2, MPI_DOUBLE, c, n2, MPI_DOUBLE, master, \
		      MPI_COMM_WORLD);
    
    //Print the computing time and result as a matrix
    if (my_rank == master) {
      t_finish = MPI_Wtime();
      printf("Time elapsed for parallel computation (s) %f \n", \
	     t_finish - t_start);

      for(i = 0; i < nr_a; i++) {
	for(j = 0; j < nc_b; j++) {
	  printf("%i\t", c[i][j]);
	}
	printf("\n");
      }
    }

    ierr = MPI_Finalize();

}
